### ISSUE TITLE
ImportControl can use the wrong package authorization if two packages start identically

### ISSUE THREAD

**Issue by ncorai** [checkstyle/checkstyle#3409]
... and the package with the shortest name is defined first in `import-control.xml`.

For instance, given

``` XML
<import-control pkg="root">
    <subpackage name="foo">
        <allow pkg="root.foo2"/>
    </subpackage>
    <subpackage name="foobar">
        <allow pkg="root.foobar2"/>
    </subpackage>
</import-control>
```

class `root.foobar.Baz` will get a warning when importing `root.foobar2` but not when importing `root.foo2`

Based on a cursory look at the code base, I believe the bug happens when attempting to retrieve the authorizations for `root.foobar.Baz`:

Method `com.puppycrawl.tools.checkstyle.checks.imports.ImportControlCheck#visitToken` will call `com.puppycrawl.tools.checkstyle.checks.imports.PkgControl#locateFinest` on the root node, then iterate its children until it finds one that the package of the inspected class _starts with_ (culprit is `if (forPkg.startsWith(fullPackage)) {` - `PkgControl.java#82` as of version `8ad9441`).

In our example, `root.foo` is defined before `root.foobar` so it is tested first and `Baz`'s package `root.foobar` does start with `root.foo`, thus `foo`'s authorizations are used instead.
##

**Comment by ncorai** [comment]
Bump!

Note: the bug report not only describes the issue but also provides the solution to fix it, which shouldn't take long.

**Comment by romani** [comment]
sounds good :) , but please follow demands on issue reporting - http://checkstyle.sourceforge.net/report_issue.html#How_to_report_a_bug
to let me approve/confirm the issue.

if fix is simple - please send PR.

**Comment by jochenvdv** [comment]
I tried reproducing this issue by doing exactly as @ncorai said, but I get a violation only for importing `root.foo2` (as expected), not for `root.foobar2`:

    ~/checkstyle-issue-3409 $ javac root/foobar/Baz.java

    ~/checkstyle-issue-3409 $ tree
    .
    ├── checkstyle-7.2-all.jar
    ├── config.xml
    ├── import-control.xml
    └── root
        ├── foo
        │   └── package-info.java
        ├── foo2
        │   └── package-info.java
        ├── foobar
        │   ├── Baz.class
        │   └── Baz.java
        └── foobar2
            └── package-info.java

    5 directories, 8 files

    ~/checkstyle-issue-3409 $ cat config.xml
    <?xml version="1.0"?>
    <!DOCTYPE module PUBLIC
            "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
            "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
    <module name="Checker">
        <module name="TreeWalker">
            <module name="ImportControl">
                <property name="file" value="import-control.xml"/>
            </module>
        </module>
    </module>

    ~/checkstyle-issue-3409 $ cat import-control.xml
    <?xml version="1.0"?>
    <!DOCTYPE import-control PUBLIC
        "-//Puppy Crawl//DTD Import Control 1.2//EN"
        "http://checkstyle.sourceforge.net/dtds/import_control_1_2.dtd">
    <import-control pkg="root">
        <subpackage name="foo">
            <allow pkg="root.foo2"/>
        </subpackage>
        <subpackage name="foobar">
            <allow pkg="root.foobar2" />
        </subpackage>
    </import-control>

    ~/checkstyle-issue-3409 $ cat root/foo/package-info.java
    package root.foo;

    ~/checkstyle-issue-3409 $ cat root/foo2/package-info.java
    package root.foo2;

    ~/checkstyle-issue-3409 $ cat root/foobar/Baz.java
    package root.foobar;

    import root.foobar2.*;
    import root.foo2.*;

    public class Baz {}

    ~/checkstyle-issue-3409 $ cat root/foobar2/package-info.java
    package root.foo2;

    ~/checkstyle-issue-3409 $ java -jar checkstyle-7.2-all.jar -c config.xml root/foobar/Baz.java
    Starting audit...
    [ERROR] /home/jochenvdv/checkstyle-issue-3409/root/foobar/Baz.java:4:1: Disallowed import - root.foo2.*. [ImportControl]
    Audit done.
    Checkstyle ends with 1 errors.

**Comment by romani** [comment]
if author did not come back to fix description as requests with all details , we will close the issue. 
Lets say dead line is 30 Dec 2016.

**Comment by ncorai** [comment]
Maybe the issue got fixed in a more recent version. I don't have too much time right now (and I don't have 7.2 installed) but I'll give it another go in a couple of weeks.

**Comment by romani** [comment]
It is 1 min to download http://checkstyle.sourceforge.net/cmdline.html#Download_and_Run and few min to launch on your code with your config. Let me know if if have priblems with this.

**Comment by vboerchers** [comment]
As far as I remember I fixed this problem with #2999 as I re-wrote that part of the file matcher. A commit message described that but due to pull request policy all commits had to be squashed into one.

**Comment by romani** [comment]
Thanks to all , I am closing this issue.
